packages/back-end/src/scim/users/createUser.ts

```diff
@@ -0,0 +1,95 @@
+import { Response } from "express";
+import {
+  convertUserToManagedByIdp,
+  createUser as createNewUser,
+  getUserByEmail,
+} from "../../services/users";
+import { addMemberToOrg } from "../../services/organizations";
+import { OrganizationInterface } from "../../../types/organization";
+import { ScimUserPutOrPostRequest } from "../../../types/scim";
+
+export async function createUser(req: ScimUserPutOrPostRequest, res: Response) {
+  const { externalId, name, displayName, userName } = req.body;
+
+  const org: OrganizationInterface = req.organization;
+
+  const role = org.settings?.defaultRole?.role || "readonly";
+
+  try {
+    // Look up the user in Mongo
+    let user = await getUserByEmail(userName);
+
+    // If the user already exists in the org, return an error
+    if (user && org.members.find((member) => member.id === user?.id)) {
+      // Check if they have an externalId and/or if they're managed by an external IDP
+      if (user.managedByIdp) {
+        // If so, return an error
+        // otherwise, update the user with the externalId and managedByIdp: true
+        return res.status(409).json({
+          schemas: ["urn:ietf:params:scim:api:messages:2.0:Error"],
+          scimType: "uniqueness",
+          detail: "User already exists in this organization",
+          status: 409,
+        });
+      } else {
+        await convertUserToManagedByIdp(
+          user.id,
+          externalId ? externalId : undefined
+        );
+      }
+    }
+
+    if (!user) {
+      // If we still can't find the user, create it
+      user = await createNewUser(
+        displayName,
+        userName,
+        "12345678", // TODO: SSO shouldn't need a password. figure out how to test this
+        false, // TODO: Double check this logic
+        true,
+        externalId ? externalId : undefined
+      );
+    }
+
+    await addMemberToOrg({
+      organization: org,
+      userId: user.id,
+      role,
+      limitAccessByEnvironment: false,
+      environments: [],
+      projectRoles: [],
+    });
+
```

Comment:
Should we introduce the pattern previously discussed where we create the response object as a clone of the request object, and only change the params necessary?

```ts
const responseObj = cloneDeep(requestBody);

responseObj.id = user.id;

return res.status(201).json(responseObj);
```

I'm happy to make that change on all of the endpoints (where it makes sense).

Reply:
I think this makes sense since the shape of the object will depend on the idP and what they decide to send us. I can make this quick change